clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -0,0 +1,660 @@
+}
+
+template <>
+memref_t
+scheduler_tmpl_t<memref_t, reader_t>::create_region_separator_marker(memref_tid_t tid,
+                                                                     uintptr_t value)
+{
+    memref_t record = {};
+    record.marker.type = TRACE_TYPE_MARKER;
+    record.marker.marker_type = TRACE_MARKER_TYPE_WINDOW_ID;
+    record.marker.marker_value = value;
+    record.marker.tid = tid;
+    return record;
+}
+
+template <>
+void
+scheduler_tmpl_t<memref_t, reader_t>::print_record(const memref_t &record)
+{
+    fprintf(stderr, "tid=%" PRId64 " type=%d", record.instr.tid, record.instr.type);
+    if (type_is_instr(record.instr.type))
+        fprintf(stderr, " pc=0x%zx size=%zu", record.instr.addr, record.instr.size);
+    fprintf(stderr, "\n");
+}
+
+/******************************************************************************
+ * Specializations for scheduler_tmpl_t<record_reader_t>, aka record_scheduler_t.
+ */
+
+template <>
+std::unique_ptr<dynamorio::drmemtrace::record_reader_t>
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::get_default_reader()
+{
+    return std::unique_ptr<default_record_file_reader_t>(
+        new default_record_file_reader_t());
+}
+
+template <>
+std::unique_ptr<dynamorio::drmemtrace::record_reader_t>
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::get_reader(const std::string &path,
+                                                             int verbosity)
+{
+    // TODO i#5675: Add support for other file formats, particularly
+    // .zip files.
+    return std::unique_ptr<dynamorio::drmemtrace::record_reader_t>(
+        new default_record_file_reader_t(path, verbosity));
+}
+
+template <>
+bool
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::record_type_has_tid(
+    trace_entry_t record, memref_tid_t &tid)
+{
+    if (record.type != TRACE_TYPE_THREAD)
+        return false;
+    tid = static_cast<memref_tid_t>(record.addr);
+    return true;
+}
+
+template <>
+bool
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::record_type_is_instr(
+    trace_entry_t record)
+{
+    return type_is_instr(static_cast<trace_type_t>(record.type));
+}
+
+template <>
+bool
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::record_type_is_marker(
+    trace_entry_t record, trace_marker_type_t &type, uintptr_t &value)
+{
+    if (record.type != TRACE_TYPE_MARKER)
+        return false;
+    type = static_cast<trace_marker_type_t>(record.size);
+    value = record.addr;
+    return true;
+}
+
+template <>
+trace_entry_t
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::create_region_separator_marker(
+    memref_tid_t tid, uintptr_t value)
+{
+    // We ignore the tid.
+    trace_entry_t record;
+    record.type = TRACE_TYPE_MARKER;
+    record.size = TRACE_MARKER_TYPE_WINDOW_ID;
+    record.addr = value;
+    return record;
+}
+
+template <>
+void
+scheduler_tmpl_t<trace_entry_t, record_reader_t>::print_record(
+    const trace_entry_t &record)
+{
+    fprintf(stderr, "type=%d\n", record.type);
```

Comment:
Printing just the type here?

Reply:
Hadn't needed trace_entry_t diagnostics yet.  But easy to add size+addr so added.